stratosphere/pm/source/boot2/boot2_api.cpp

```diff
@@ -0,0 +1,294 @@
+            ncm::TitleId{TitleId_Fatal},       /* fatal */
+            ncm::TitleId{TitleId_Ro},          /* ro */
+            ncm::TitleId{TitleId_Profiler},    /* profiler */
+            ncm::TitleId{TitleId_Sdb},         /* sdb */
+            ncm::TitleId{TitleId_Migration},   /* migration */
+            ncm::TitleId{TitleId_Grc},         /* grc */
+            ncm::TitleId{TitleId_Olsc},        /* olsc */
+        };
+        constexpr size_t NumAdditionalMaintenanceLaunchPrograms = sizeof(AdditionalMaintenanceLaunchPrograms) / sizeof(AdditionalMaintenanceLaunchPrograms[0]);
+
+        /* Helpers. */
+        inline bool IsHexadecimal(const char *str) {
+            while (*str) {
+                if (!std::isxdigit(static_cast<unsigned char>(*(str++)))) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        void LaunchTitle(u64 *out_process_id, const ncm::TitleLocation &loc, u32 launch_flags) {
+            u64 process_id = 0;
+
+            /* Don't launch a title twice during boot2. */
+            if (pm::info::HasLaunchedTitle(loc.title_id)) {
+                return;
+            }
+
+            switch (pm::shell::LaunchTitle(&process_id, loc, launch_flags)) {
+                case ResultKernelResourceExhausted:
+                    /* Out of resource! */
+                    std::abort();
+                case ResultKernelOutOfMemory:
+                    /* Out of memory! */
+                    std::abort();
+                case ResultKernelLimitReached:
+                    /* Limit Reached! */
+                    std::abort();
+                default:
+                    /* We don't care about other issues. */
+                    break;
+            }
+
+            if (out_process_id) {
+                *out_process_id = process_id;
+            }
+        }
+
+        void LaunchList(const ncm::TitleId *launch_list, size_t num_entries) {
+            for (size_t i = 0; i < num_entries; i++) {
+                LaunchTitle(nullptr, ncm::MakeTitleLocation(launch_list[i], ncm::StorageId::NandSystem), 0);
+            }
+        }
+
+        bool GetGpioPadLow(GpioPadName pad) {
+            GpioPadSession button;
+            if (R_FAILED(gpioOpenSession(&button, pad))) {
+                return false;
+            }
+
+            /* Ensure we close even on early return. */
+            ON_SCOPE_EXIT { gpioPadClose(&button); };
+
+            /* Set direction input. */
+            gpioPadSetDirection(&button, GpioDirection_Input);
+
+            GpioValue val;
+            return R_SUCCEEDED(gpioPadGetValue(&button, &val)) && val == GpioValue_Low;
+        }
+
+        bool IsMaintenanceMode() {
+            /* Contact set:sys, retrieve boot!force_maintenance. */
+            DoWithSmSession([&]() {
+                R_ASSERT(setsysInitialize());
+            });
+            {
+                ON_SCOPE_EXIT { setsysExit(); };
```

Comment:
Having separate scopes and separate `DoWithSmSession` blocks for `set:sys` and `gpio` seem to me to be more harmful to readability than they are beneficial, unless the idea is to make it to gpio init errors wouldn't prevent maintenance forced by `set:sys`. The `DoWithSmSession`/`ON_SCOPE_EXIT` pattern doesn't seem to be especially conducive to initializing multiple sessions safely though... here's an idea:
```cpp
template<Result Initializer(), void Finalizer()>
class ServiceHolder {
  public:
    // non-copyable boilerplate
    // move semantics boilerplate
    Result Initialize() {
        if(has_initialized) {
            std::abort(); // sanity
        }
        initialization_result = Initializer();
        if(R_OK(initialization_result)) {
            has_initialized = true;
        }
        return initialization_result; // for convenient R_ASSERT composition
    }
    void Finalize() {
        if(!has_initialized) {
            std::abort(); // invalid state
        }
        Finalizer();
        has_initialized = false;
    }
    void TryFinalize() {
        if(has_initialized) {
            Finalizer();
        }
        has_initialized = false;
    }
    explicit operator bool() const {
        return has_initialized;
    }
    Result GetResult() const {
        return initialization_result;
    }
    ~ServiceHolder() {
        TryFinalize();
    }
  private:
    Result initialization_result;
    bool has_initialized = false;
};
```
````cpp
ServiceHolder<setsysInitialize, setsysExit> setsys_holder;
ServiceHolder<gpioInitialize, gpioExit> gpio_holder;
DoWithSmSession([&]() {
    R_ASSERT(setsys_holder.Initialize());
    R_ASSERT(gpio_holder.Initialize());
});
// check set:sys
// check gpio
````
or,
```cpp
ServiceHolder<setsysInitialize, setsysExit> setsys_holder;
ServiceHolder<gpioInitialize, gpioExit> gpio_holder;
DoWithSmSession([&]() {
    setsys_holder.Initialize();
    gpio_holder.Initialize();
});
R_ASSERT(setsys_holder.GetResult());
// check set:sys, potentially return early and skip second R_ASSERT
R_ASSERT(gpio_holder.GetResult());
// check gpio
```

Reply:
I love this idea. However, let's do it like AutoCloseMap -- constructor sets this->result = Initialize(). Destructor closes if Result was success -- don't require member calls to init, so that object scope is directly tied to service state.